src/exec/use_pty/pipe.rs

```diff
@@ -222,15 +360,70 @@ impl<R: Read, W: Write> Buffer<R, W> {
-        self.start = 0;
-        self.end = 0;
+        // We need to find all the sub-buffers with data.
+        if self.head < self.tail {
+            // If head is less than tail, there is a single sub-buffer with data that extends from
+            // head to one position before tail.
+            //         ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │   │ M │ D │ D │ D │   │   │   │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲           ▲
+            //           │           │
+            //         head         tail
+            // In the extreme case where head is as close as possible to tail, the sub-buffer will have a
+            // length of one. We can be sure that head and tail are not equal because the buffer is
+            // not empty.
+            //         ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │   │ M │ D │   │   │   │   │   │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲   ▲
+            //           │   │
+            //         head  tail
+            write.write_all(&self.buffer[self.head..self.tail])?;
+        } else {
+            // If head is greater than tail, there are two sub-buffers with data. The first one
+            // extends from head to the end of the buffer. and the second one goes from the start
+            // of the buffer to one position before tail.
+            // ┌───┐               ┌───────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │   │   │   │ M │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //       ▲               ▲
+            //       │               │
+            //     tail             head
+            //
+            // In the extreme case where head is the last position, the first sub-buffer will have
+            // a length of one.
+            // ┌───────┐                   ┌───┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │ D │   │   │   │   │ M │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲                   ▲
+            //           │                   │
+            //         tail                 head
+            // On the other hand if tail is as close as it can be to head, the buffer is full
+            // ┌───────┐   ┌───────────────────┐
+            // ┌───┬───┬───┬───┬───┬───┬───┬───┐
+            // │ D │ D │ M │ D │ D │ D │ D │ D │
+            // └───┴───┴───┴───┴───┴───┴───┴───┘
+            //           ▲   ▲
+            //           │   │
+            //         tail head
+            // We know they cannot be equal because the buffer is not empty.
+            write.write_all(&self.buffer[self.head..])?;
+            write.write_all(&self.buffer[..self.tail])?;
+        };
+
+        // Now that we have written all the data from the buffer. We can mark it as empty
```

Comment:
I don't think this is necessary. so long the ring buffer is correctly implemented (and used) it should be fine to let the cursors keep growing. `self.head == self.tail` is the empty condition after all; the cursors don't necessarily have to be equal to `0`.